### PR TITLE
[SPARK-15688][SQL] RelationalGroupedDataset.toDF should not add group by xpressions that are already added in the aggregate expressions

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?

The dataset aggregate methods (agg) computes aggregates and produces
the resulting dataset. Currently the resulting data set includes the 
grouping columns even though the aggregation expression already
contains the columns resulting in duplicate columns.

``` SQL
val df = Seq(1 -> "a", 2 -> "b", 3 -> "b").toDF("col1", "col2")
val df1 = df.groupBy("col1").agg($"col1", count("*")).show()
+----+----+--------+
|col1|col1|count(1)|
+----+----+--------+
|   3|   3|       1|
|   1|   1|       1|
|   2|   2|       1|
+----+----+--------+
```

This PR addresses the problem by only including grouping columns
in the dataset if they were not already contained by the aggregate
expressions.

``` SQL
val df1 = df.groupBy("col1").agg($"col1", count("*")).show()
+----+--------+
|col1|count(1)|
+----+--------+
|   3|       1|
|   1|       1|
|   2|       1|
+----+--------+
```

**Normal Aggregate plan.**

``` SQL
Aggregate [col1#5], [col1#5,count(1) AS count(1)#15L]
+- Project [_1#2 AS col1#5,_2#3 AS col2#6]
   +- LocalRelation [_1#2,_2#3], [[0,1,1800000001,61],[0,2,1800000001,62],[0,3,1800000001,62]]
```

**Plan with window expressions in aggregate.**

``` SQL
Project [sum(earnings)#21,grouping_id(course, year)#22,RANK() OVER (PARTITION BY grouping_id(course, year) ORDER BY sum(earnings) ASC UnspecifiedFrame)#23]
+- Project [sum(earnings)#21,grouping_id(course, year)#22,_w0#24,_w1#25,RANK() OVER (PARTITION BY grouping_id(course, year) ORDER BY sum(earnings) ASC UnspecifiedFrame)#23,RANK() OVER (PARTITION BY grouping_id(course, year) ORDER BY sum(earnings) ASC UnspecifiedFrame)#23]
   +- Window [rank(_w1#25) windowspecdefinition(_w0#24, _w1#25 ASC, ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS RANK() OVER (PARTITION BY grouping_id(course, year) ORDER BY sum(earnings) ASC UnspecifiedFrame)#23], [_w0#24], [_w1#25 ASC]
      +- !Aggregate [course#4,year#5], [sum(earnings#6) AS sum(earnings)#21,grouping_id(course#4, year#5) AS grouping_id(course, year)#22,grouping_id(course#4, year#5) AS _w0#24,sum(earnings#6) AS _w1#25]
         +- LogicalRDD [course#4,year#5,earnings#6], MapPartitionsRDD[1] at apply at Transformer.scala:22

```
## How was this patch tested?

Added tests in DatasetAggregatorSuite

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala [hunk 1]
```diff
@@ -224,6 +224,21 @@ class DatasetAggregatorSuite extends QueryTest with SharedSQLContext {
...
 
+  test("SPARK-15688: Remove redundant expressions from aggregate") {
+    val df = Seq(1 -> "a", 2 -> "b", 3 -> "b").toDF("col1", "col2")
+    val df1 = df.groupBy("col1").agg($"col1", count("*"))
+    assert(df1.schema.map(_.name) === Seq("col1", "count(1)"))
+    val df2 = df.groupBy("col1", "col2").agg($"col1", count("*"))
+    assert(df2.schema.map(_.name) === Seq("col2", "col1", "count(1)"))
+    val df3 = df.groupBy(expr("col1 + 2")).agg(expr("col1 + 2"), count("*"))
+    assert(df3.schema.map(_.name) === Seq("(col1 + 2)", "count(1)"))
+    val df4 = df.groupBy("col1").agg(min("col1"), max("col1"))
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala]: Could you put some comments to show the goal of each test cases? The comments can help reviewers understand the coverage of test cases. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala [hunk 2]
```diff
@@ -46,7 +46,18 @@ class RelationalGroupedDataset protected[sql](
 
   private[this] def toDF(aggExprs: Seq[Expression]): DataFrame = {
     val aggregates = if (df.sparkSession.sessionState.conf.dataFrameRetainGroupColumns) {
-      groupingExprs ++ aggExprs
+      val agg = Aggregate(groupingExprs, aggExprs.map(alias), df.logicalPlan)
+      val plan = df.sparkSession.sessionState.executePlan(agg).analyzed
+      val aggAnalyzed = plan.find(_.isInstanceOf[Aggregate]).getOrElse(agg).asInstanceOf[Aggregate]
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: The same here. Can you put a comment to explain the top level node might not be an `Aggregate`

**Comment by gatorsmile** [comment]
Please update the PR description and post the analyzed plan. Also added another example to show the root node is `Window`. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala (hunk 1 above)
Comment (dilipbiswal) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala]: Sure. Will do.

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala (hunk 2 above)
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: Ok.

**Comment by dilipbiswal** [comment]
@gatorsmile added comments and also updated the PR description.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala [hunk 3]
```diff
@@ -224,6 +224,26 @@ class DatasetAggregatorSuite extends QueryTest with SharedSQLContext {
...
+    // Aggregate expression contains the group by column and should appear once in the output.
+    val df1 = df.groupBy("col1").agg($"col1", count("*"))
+    assert(df1.schema.map(_.name) === Seq("col1", "count(1)"))
+    // Only the column that is missing in aggregate expression should be added to aggregate expr.
+    val df2 = df.groupBy("col1", "col2").agg($"col1", count("*"))
+    assert(df2.schema.map(_.name) === Seq("col2", "col1", "count(1)"))
+    // Complex expression in group by and aggregate clause.
+    val df3 = df.groupBy(expr("col1 + 2")).agg(expr("col1 + 2"), count("*"))
+    assert(df3.schema.map(_.name) === Seq("(col1 + 2)", "count(1)"))
+    // Group by column should be added as its not part of aggregate expression.
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala]: `Group by column` -> `Group-by expressions`. `its` -> `they are`

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala [hunk 4]
```diff
@@ -224,6 +224,26 @@ class DatasetAggregatorSuite extends QueryTest with SharedSQLContext {
     checkAnswer(df.groupBy($"j").agg(RowAgg.toColumn), Row("a", 1) :: Row("b", 5) :: Nil)
   }
 
+  test("SPARK-15688: Remove redundant expressions from aggregate") {
+    val df = Seq(1 -> "a", 2 -> "b", 3 -> "b").toDF("col1", "col2")
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala]: Could you use `WithSQLConf` here? This is to ensure this test case will not be affected by the others or the future change in the default value of `SQLConf.DATAFRAME_RETAIN_GROUP_COLUMNS.key`

Please also fix the related one in `DataFrameAggregateSuite` by adding `WithSQLConf`:  https://github.com/apache/spark/blob/bb1362eb3b36b553dca246b95f59ba7fd8adcc8a/sql/core/src/test/scala/org/apache/spark/sql/DataFrameAggregateSuite.scala#L210-L215

**Comment by gatorsmile** [comment]
LGTM except a few comments in test cases. : )

@cloud-fan @yhuai Could you please review this PR? Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala (hunk 4 above)
Comment (dilipbiswal) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala]: @gatorsmile Ah.. good idea. I will make the change. Thanks :-)

**Comment by viirya** [comment]
As the setting `spark.sql.retainGroupColumns` is set to true by default, `GroupedData.agg` will retain group by columns. So the problem is you explicitly ask the column `col1` when calling `agg` method. That is like when users explicitly do `SELECT a, a`, but we automatically convert it to `SELECT a`. Looks unreasonable to me.

**Comment by gatorsmile** [comment]
@viirya Based on my understanding, we want to avoid having duplicate output columns generated by our APIs. If users want to explicitly have two duplicate output columns, they can specify them in `agg`. That  For example, 
```Scala
df.groupBy("col1").agg($"col1", $"col1", count("*"))  
```

In addition, `spark.sql.retainGroupColumns` is an internal parameter. The default value is `true`. That means, users might not realize they can disable it. Thus, it makes more sense to avoid generating duplicate columns that are caused by retaining group-by expressions.

**Comment by dilipbiswal** [comment]
@gatorsmile Thanks a lot for your explanation. I agree that if we internally generate the aggregate expressions that is different than user specified one , then we should eliminate the duplicate expressions as opposed to simply adding the the grouping expressions like we do today.

**Comment by cloud-fan** [comment]
I don't think it's a big problem. After users understand the semantic, they can easily work around it, by not putting grouping columns in agg list, or turning off the config.
Well, it looks better if we can deduplicate it automatically, but seems it's not that easy, according to this PR. Maybe it doesn't worth.

cc @marmbrus

**Comment by viirya** [comment]
@gatorsmile this automatically deduplicate the group by columns will cause confusion. Using your example:

    df.groupBy("col1").agg(count("*"))

When users try the above API call, they will find that `groupBy` will retain group by columns. Then as they try:

    df.groupBy("col1").agg($"col1", count("*"))  

They will figure out that adding group by columns will make duplicate ones.

However, if we automatically deduplicate them. The users first think that `groupBy` will include group by columns. Then when adding group by columns explicitly, the duplicate group by columns are not shown up. But as they add more, the group by columns show up again...

Personally I think the behavior above is inconsistent.

**Comment by gatorsmile** [comment]
@viirya This is a design decision. So far, both ways are not perfect. 

In my mind, we have to consider the use cases here. If users want to have duplicate columns, they should not use the duplicate names. Do you think this makes sense? 

That means, we should not remove the duplicate in the following scenario:
```
df.groupBy("col1").agg($"col1".as("col1_replica"), count("*"))

The expected output schema:
+----+------------+--------+
|col1|col1_replica|count(1)|
+----+------------+--------+
```
If they do not change the column name, I am unable to find any usage senario for duplicating the columns. 
```
df.groupBy("col1").agg($"col1", count("*"))
df.groupBy("col1").agg(count("*"))

The expected output schema of the above two:
+----+--------+
|col1|count(1)|
+----+--------+
```

What is your opinions?

**Comment by viirya** [comment]
@gatorsmile Your last example just shows the inconsistency. Given two different parameters, `$"col1", count("*")` and `count("*")`, you get the same output. I think this confuses users. In contrast, original API usage has consistent behavior: group by columns are retained, and when you add group by columns in `agg` API call, it means you want to add duplicate ones.

Of course, that is my opinion. I am not against this change if others think it is ok.

**Comment by gatorsmile** [comment]
@viirya I am just wondering why users need a dataframe with duplicate column names? Could you give me a usage scenario?

**Comment by viirya** [comment]
@gatorsmile I am not stand for it due to actual use cases, but the API behavior consistency. If we disallow duplicate group by columns, then we should do filtering them all. Let

    df.groupBy("col1").agg($"col1", $"col1", count("*"))

show duplicate group by columns but disallow

    df.groupBy("col1").agg($"col1", count("*"))

looks confusing...

**Comment by gatorsmile** [comment]
@viirya You know, as I said above, both ways are not perfect. My inputs are just based on my previous design experiences. All the design decisions I made are based on usage scenarios. 

This PR is to refine the behavior of `spark.sql.retainGroupColumns`. I think that might be why @yhuai opened this JIRA. 

The problem you mentioned above is another issue. I think we should not deduplicating the columns, if users really want to do something absurd. : )

**Comment by marmbrus** [comment]
I haven't looked at this PR in detail, but I'm the originator of this suggestion.  It seemed odd to me that we were auto adding columns when I was already specifying them.  I agree if the user manually specifies a column twice, then it needs to show up twice.

Additionally, we are now too late in the 2.0 process to make changes to the semantics.  So its probably okay to just close this as `wont fix`.

**Comment by dilipbiswal** [comment]
@marmbrus Thanks Michael. I am going to close this.

**Comment by gatorsmile** [comment]
Thank you @marmbrus @dilipbiswal and @viirya !

**Pull request closed without merging**